### PR TITLE
FOGL-2273 Notifications GUI

### PR DESCRIPTION
**Pull request by mshariq-nerd**
Functionality Integrated in this Pull request:

1) Add notification service 
2) Add notification instance via notification wizard
3) Update  notification instance configuration using modal 
4) Delete notification instance
5) Refresh (synch) of notification list

### CODE REVIEW

File: src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.ts [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+        previousButton.textContent = 'Previous';
+
+        // To verify if service with given name already exist
+        // const isServiceNameExist = false;
+        // this.schedulesName.some(item => {
+        //   return formValues['name'].trim() === item.name;
+        // });
+        // if (isServiceNameExist) {
+        //   this.alertService.error('A south service or north task instance already exists with this name.');
+        //   return false;
```
Comment (praveen-garg) [inline src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.ts]: remove these comments

File: src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.ts [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+        nxtButton.disabled = false;
+        break;
+      default:
+        break;
+    }
+  }
+
+  moveNext() {
+    this.isRulePlugin = true;
+    this.isValidName = true;
```
Comment (praveen-garg) [inline src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.ts]: move this statement before `this.isRulePlugin = true;`

File: src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.ts [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+        }
+        if (formValues['delivery'].length > 1) {
+          this.isSinglePlugin = false;
+          return;
+        }
+
+        if (formValues['delivery'].length > 0) {
+          this.payload.channel = formValues['delivery'][0];
+        }
+        // create payload to pass in add service
```
Comment (praveen-garg) [inline src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.ts]: remove this comment

File: src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.ts [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+   */
+  public addNotificationInstance(payload: any) {
+    /** request started */
+    this.ngProgress.start();
+    this.notificationService.addNotificationInstance(payload)
+      .subscribe(
+        () => {
+          /** request done */
+          this.ngProgress.done();
+          this.alertService.success('Notification service added successfully.', true);
```
Comment (praveen-garg) [inline src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.ts]: what is the API response here. can not we use that as is. if that is not that good, better to say `'Notification instance <name> added successfully.`

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  public useProxy: 'true';
+  public category: any;
+  public ruleConfiguration: any;
+  public deliveryConfiguration: any;
+  public notificationRecord: any;
+  public changedChildConfig = [];
+
+  rulePluginChangedConfig = [];
+  deliveryPluginChangedConfig = [];
+  notificationChangedConfig = [];
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]: on modal, even if no config is changed and Save is clicked. old request is being repeated.

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    if (activeFilterTab !== undefined) {
+      const activeContentBody = <HTMLElement>activeFilterTab.getElementsByClassName('card-content')[0];
+      activeFilterTab.classList.remove('is-active');
+      activeContentBody.hidden = true;
+    }
+
+    const modalWindow = <HTMLDivElement>document.getElementById('south-service-modal');
+    if (isOpen) {
+      // this.notify.emit(false);
+      // this.svcCheckbox.setValue(this.service['schedule_enabled']);
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]: remove these comments?

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+      modalWindow.classList.add('is-active');
+      return;
+    }
+    // this.notify.emit(false);
+    // this.isAdvanceConfig = true;
+    // this.getAdvanceConfig(null);
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]: remove these comments?

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    modalWindow.classList.remove('is-active');
+  }
+
+  public getRuleConfiguration(): void {
+    const categoryValues = [];
+    const notificationName = this.notification['name'].substr(this.notification['name'].indexOf(' ') + 1);
+    this.configService.getCategory(`rule${notificationName}`).
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]: this is cool! for string interpolation!

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      subscribe(
+        (data) => {
+          if (!isEmpty(data)) {
+            categoryValues.push(data);
+            this.ruleConfiguration = { key: `rule${notificationName}`, value: categoryValues };
+            this.useRuleProxy = 'true';
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]: do we set to false? I think its always true. Or we change it based on config data chnage for Rule

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    this.ngProgress.start();
+    const categoryValues = [];
+    const notificationName = this.notification['name'].substr(this.notification['name'].indexOf(' ') + 1);
+    this.configService.getCategory(`delivery${notificationName}`).
+      subscribe(
+        (data) => {
+          if (!isEmpty(data)) {
+            categoryValues.push(data);
+            this.deliveryConfiguration = { key: `delivery${notificationName}`, value: categoryValues };
+            this.useDeliveryProxy = 'true';
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]: do we set to false? I think its always true. Or we change it based on config data chnage for Delivery Channel

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+
+  /**
+   * Open delete modal
+   * @param message   message to show on alert
+   * @param action here action is 'deleteNotification'
+   */
+  openDeleteModal(name: string) {
+    this.notificationRecord = {
+      name: name,
+      message: 'Warning: Deleting this notification can not be undone. Continue',
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]:
```suggestion
      message: 'Deleting this notification can not be undone. Continue?',
```

File: src/app/components/core/notifications/notifications.component.html [hunk 12]
```diff
@@ -0,0 +1,80 @@
+<div class="container is-fluid">
+  <div>
+    <p class="notification is-warning" *ngIf="!isNotificationServiceAvailable">
+      Notification service is not configured
+      and enabled. <a (click)="addNotificationService()"> Enable Now</a></p>
+    <p class="notification is-warning" *ngIf="isNotificationServiceAvailable && !isNotificationServiceEnabled">
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notifications.component.html]: https://bulma.io/documentation/components/message/#message-body-only

File: src/app/components/core/notifications/notifications.component.html [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      Notification service is not enabled. <a (click)="enableNotificationService()"> Enable Now</a></p>
+  </div>
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notifications.component.html]: We should also provide an option (message body info) to allow to `Disable Now` (to disable notification schedule) and show a warning alert to confirm .

Not sure if we should also have  option to `Delete` as well?!

**Comment by praveen-garg** [comment]
Adding `NF #1` appears as `Notification NF %231`

`Notification ` prefix is a known API issue FOGL-2645. I think we also need to get this encoding issue  reported / fixed.

File: src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.html [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    </div>
+    <div class="step-item" id="4">
+      <div class="step-marker">3</div>
+      <div class="step-details">
+        <p>Delivery Channel</p>
+      </div>
+    </div>
+    <div hidden class="step-item" id="5">
+      <div class="step-details">
+        <p>Review Delivery Configuration</p>
```
Comment (praveen-garg) [inline src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.html]: <p>Review Delivery Channel Configuration</p>

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+      for (const e of <any>el) {
+        e.click();
+      }
+    }
+    if (this.useDeliveryProxy) {
+      const ele = <HTMLCollection>document.getElementsByClassName('vci-proxy-delivery');
+      for (const e of <any>ele) {
+        e.click();
+      }
+      this.toggleModal(false);
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]: should not this be out of `if (this.useDeliveryProxy)` block?

File: src/app/components/core/configuration-manager/configuration-manager.component.ts [hunk 16]
```diff
@@ -52,8 +52,10 @@ export class ConfigurationManagerComponent implements OnInit {
                 description: element.description
               });
             }
-            this.rootCategories = this.rootCategories.filter(el => el.key.toUpperCase() !== 'SOUTH');
-            this.rootCategories = this.rootCategories.filter(el => el.key.toUpperCase() !== 'NORTH');
+            this.rootCategories = this.rootCategories.filter(el => el.key.toUpperCase() !== 'SOUTH')
+            .filter(el => el.key.toUpperCase() !== 'NORTH')
+            .filter(el => el.key.toUpperCase() !== 'NOTIFICATIONS')
+            .filter(el => el.key.toUpperCase() !== 'FOGLAMP NOTIFICATION');
```
Comment (praveen-garg) [inline src/app/components/core/configuration-manager/configuration-manager.component.ts]: did we never dropped for Filters?

File: src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.html [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        <div class="box step-content" id="c-5">
+          <app-view-config-item *ngIf="pageId == 4"
+            [categoryConfigurationData]="deliveryPluginConfigurationData" [useProxy]="useProxy"
+            (onConfigChanged)="getPluginChangedConfig($event, deliveryPluginConfigurationData, pageId)" [pageId]="'wizard'">
+          </app-view-config-item>
+        </div>
+        <div class="box step-content" id="c-6">
+          <div class="field is-horizontal">
+            <div class="field-label">
+              <label class="label">Trigger Notification</label>
```
Comment (praveen-garg) [inline src/app/components/core/notifications/add-notification-wizard/add-notification-wizard.component.html]: appearing as 

`Trigger` 
`Notification`

File: src/app/components/core/notifications/notification-modal/notification-modal.component.ts [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+
+  /**
+   * Open delete modal
+   * @param message   message to show on alert
+   * @param action here action is 'deleteNotification'
+   */
+  openDeleteModal(name: string) {
+    this.notificationRecord = {
+      name: name,
+      message: 'Deleting this notification can not be undone. Continue',
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.ts]: 'Deleting this notification instance can not be undone. Continue'

File: src/app/components/core/notifications/notifications.component.html [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+      <article class="message is-warning">
+        <div class="message-body">
+          Notification service is not enabled. <a (click)="enableNotificationService()"><strong>enable now</strong></a>
+        </div>
+      </article>
+    </ng-container>
+    <ng-container *ngIf="isNotificationServiceAvailable && isNotificationServiceEnabled">
+      <article id='messageDiv' class="message is-info">
+        <div class="message-body">
+          Notification service is configured and enabled. <a (click)="disableNotificationService()">disable now</a>
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notifications.component.html]: should give  a warning with options Cancel | Disable

File: src/app/components/core/notifications/notifications.component.ts [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+            }
+          }
+        });
+      });
+    this.getNotificationInstance();
+  }
+
+  addNotificationService() {
+    const payload = {
+      name: 'FogLAMP Notification',
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notifications.component.ts]: 'FogLAMP Notifications'

File: src/app/components/core/notifications/notifications.component.ts [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+    this.notificationModal.toggleModal(true);
+  }
+
+  addNotificationInstance() {
+    this.router.navigate(['/notification/add']);
+  }
+
+  public closeMessage(isOpen: Boolean) {
+    const modalName = <HTMLElement>document.getElementById('messageDiv');
+    console.log('modalName', modalName);
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notifications.component.ts]: please remove

File: src/app/components/core/scheduler/list-schedules/list-schedules.component.ts [hunk 22]
```diff
@@ -96,7 +98,7 @@ export class ListSchedulesComponent implements OnInit {
           //  filter by South and North categories name
           this.scheduleData = [];
           schedules.forEach(sch => {
-            if (!sn.includes(sch.name)) {
+            if (!sn.includes(sch.name) && sch.name.toUpperCase() !== 'FOGLAMP NOTIFICATION') {
```
Comment (praveen-garg) [inline src/app/components/core/scheduler/list-schedules/list-schedules.component.ts]: I think, filtering of schedules (e.g. south) is being handled at diff place

Comment (praveen-garg) [reply src/app/components/core/scheduler/list-schedules/list-schedules.component.ts]: please filter by process name `notification_c`. even for south and north, we can use filter schedules having process name [south, south_c, north, north_c] and avoid the call for categories

File: src/app/components/core/notifications/notifications.component.ts [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+
+    /** request start */
+    this.ngProgress.start();
+
+    this.servicesHealthService.addService(payload)
+      .subscribe(
+        () => {
+          /** request done */
+          this.ngProgress.done();
+          this.alertService.success('Notification service added successfully.', true);
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notifications.component.ts]: what do we get from Server API?

File: src/app/components/core/notifications/notification-modal/notification-modal.component.html [hunk 24]
```diff
@@ -0,0 +1,57 @@
+<div id="south-service-modal" class="modal">
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notification-modal/notification-modal.component.html]: id `notification-instance-modal`

File: src/app/components/core/notifications/notifications.component.ts (hunk 23 above)
Comment (mshariq-nerd) [reply src/app/components/core/notifications/notifications.component.ts]: return from server
`{"name": "FogLAMP Notifications", "id": "e8064fb9-e302-41b6-8a02-a290c41305b7"}`

File: src/app/components/core/notifications/notifications.component.ts [hunk 25]
```diff
@@ -0,0 +1,237 @@
...
+    const payload = {
+      name: this.notificationServiceName,
+      type: 'notification',
+      enabled: true
+    };
+
+    /** request start */
+    this.ngProgress.start();
+
+    this.servicesHealthService.addService(payload)
```
[the remark this answers is not in the document]
Comment (mshariq-nerd) [reply src/app/components/core/notifications/notifications.component.ts]: This is the refrence name of `ServicesHealthService` class we have.

[the remark this answers is not in the document]
Comment (praveen-garg) [reply src/app/components/core/notifications/notifications.component.ts]: Okay. we should name it better in some future refactoring PR.

File: src/app/components/core/notifications/notifications.component.ts (hunk 23 above)
Comment (praveen-garg) [reply src/app/components/core/notifications/notifications.component.ts]: Okay! then its fine. May be we would have alert `.success('<name> service added successfully.', `

File: src/app/components/core/configuration-manager/configuration-manager.component.ts [hunk 26]
```diff
@@ -52,8 +52,10 @@ export class ConfigurationManagerComponent implements OnInit {
                 description: element.description
               });
             }
-            this.rootCategories = this.rootCategories.filter(el => el.key.toUpperCase() !== 'SOUTH');
-            this.rootCategories = this.rootCategories.filter(el => el.key.toUpperCase() !== 'NORTH');
+            this.rootCategories = this.rootCategories.filter(el => el.key.toUpperCase() !== 'SOUTH')
+            .filter(el => el.key.toUpperCase() !== 'NORTH')
+            .filter(el => el.key.toUpperCase() !== 'NOTIFICATIONS')
+            .filter(el => el.key.toUpperCase() !== 'FOGLAMP NOTIFICATIONS');
```
[the remark this answers is not in the document]
Comment (mshariq-nerd) [reply src/app/components/core/configuration-manager/configuration-manager.component.ts]: filters is still there. we are not dropping them. notificationRule, notificationDelivery are the child of `notifications` category so no need to filter them separately.

File: src/app/components/core/notifications/notifications.component.html [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+              <span class="icon is-small">
+                <i class="fa fa-sync"></i>
+              </span>
+            </a>
+          </p>
+        </div>
+      </div>
+      <a id="add_notification_instance" class="fix-pad button is-light" (click)="addNotificationInstance()">Add
+        notification
+        instance &nbsp;
```
Comment (praveen-garg) [inline src/app/components/core/notifications/notifications.component.html]: `Add Notification Instance` or `Create Notification Instance`

**Pull request merged**